Simbody/src/GeneralForceSubsystem.cpp

```diff
@@ -34,12 +34,457 @@
 #include "simbody/internal/GeneralForceSubsystem.h"
 #include "simbody/internal/SimbodyMatterSubsystem.h"
 #include "simbody/internal/MultibodySystem.h"
+#include <iostream>
+#include <exception>
 
 #include "ForceImpl.h"
 
+#include <memory>
+
+//Threading constants used by CalcForcesTask
+namespace {
+const int NumNonParallelThreads = 1;
+const int NonParallelForcesIndex = 0;
+}
 
 namespace SimTK {
 
+/** Calculates each enabled force's contribution in the MultibodySystem.
+CalcForcesTask allows force calculations to occur in parallel with non-parallel
+forces being calculated on Thread NonParallelForcesIndex and all other parallel
+forces being calculated on seperate threads.*/
+class CalcForcesTask : public ParallelExecutor::Task {
+public:
+    //By default, CalcForcesTask will be in Mode All.
+    CalcForcesTask(){};
+    
+    virtual CalcForcesTask* clone() const = 0;
+    
+    virtual void initializeAll(
+            const State& s,
+            const Array_<Force*>& enabledNonParallelForces,
+            const Array_<Force*>& enabledParallelForces,
+            Vector_<SpatialVec>& rigidBodyForces,
+            Vector_<Vec3>& particleForces,
+            Vector& mobilityForces) = 0;
+    virtual void initializeCachedAndNonCached(
+            const State& s,
+            const Array_<Force*>& enabledNonParallelForces,
+            const Array_<Force*>& enabledParallelForces,
+            Vector_<SpatialVec>& rigidBodyForces,
+            Vector_<Vec3>& particleForces,
+            Vector& mobilityForces,
+            Vector_<SpatialVec>& rigidBodyForceCache,
+            Vector_<Vec3>& particleForceCache,
+            Vector& mobilityForceCache) = 0;
+    virtual void initializeNonCached(
+            const State& s,
+            const Array_<Force*>& enabledNonParallelForces,
+            const Array_<Force*>& enabledParallelForces,
+            Vector_<SpatialVec>& rigidBodyForces,
+            Vector_<Vec3>& particleForces,
+            Vector& mobilityForces) = 0;
+    virtual void initialize() = 0; 
+    virtual void execute(int threadIndex) = 0;
```

Comment:
These two are echoing pure virtuals in the base class aren't they? Better not to list them at all -- the compiler will make sure they get implemented.